### PR TITLE
SONARJAVA-1506 - Handle type of unbounded wildcard <?>

### PR DESCRIPTION
**Pull request by DidierBesset**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/resolve/Symbols.java [hunk 1]
```diff
@@ -123,6 +124,17 @@ public String toString() {
     unknownType.supertype = null;
     unknownType.interfaces = ImmutableList.of();
     unknownSymbol.type = unknownType;
+
+    anyType = new JavaType.ClassJavaType(unknownSymbol) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Symbols.java]: MAJOR Make this a named "static" inner class. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2694)

File: java-squid/src/main/java/org/sonar/java/resolve/JavaType.java [hunk 2]
```diff
@@ -384,5 +389,27 @@ public JavaType substitution(TypeVariableJavaType typeVariableType) {
...
+    }
+
+    private boolean checkWildCards(ParametrizedTypeJavaType superType) {
+      Map<String, JavaType> itsTypes = new HashMap<>();
+      for (Map.Entry<TypeVariableJavaType, JavaType> entry : superType.typeSubstitution.substitutionEntries()) {
+        itsTypes.put(entry.getKey().name(), entry.getValue());
+      }
+      for (Map.Entry<TypeVariableJavaType, JavaType> entry : typeSubstitution.substitutionEntries()) {
+        JavaType itsType = itsTypes.get(entry.getKey().name());
+        if (Symbols.anyType != itsType) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/JavaType.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-squid/src/main/java/org/sonar/java/resolve/JavaType.java [hunk 3]
```diff
@@ -228,6 +230,9 @@ public boolean isSubtypeOf(String fullyQualifiedName) {
 
     @Override
     public boolean isSubtypeOf(Type superType) {
+      if (superType == Symbols.anyType) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/JavaType.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-squid/src/main/java/org/sonar/java/resolve/JavaType.java [hunk 4]
```diff
@@ -228,6 +230,9 @@ public boolean isSubtypeOf(String fullyQualifiedName) {
 
     @Override
     public boolean isSubtypeOf(Type superType) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/JavaType.java]: MAJOR The Cyclomatic Complexity of this method "isSubtypeOf" is 12 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.

**Comment by m-g-sonar** [comment]
is reworked in another branch.

**Pull request closed without merging**